File: components/Pagination.go
package components

import (
	"fmt"

	g "github.com/maragudk/gomponents"
	c "github.com/maragudk/gomponents/components"
	h "github.com/maragudk/gomponents/html"
)

type PaginationProps struct {
	CurrentPage int
	TotalPages  int
	PageSize    int
}

func Pagination(p *PaginationProps) g.Node {
	classes := c.Classes{
		"pagination": true,
	}

	if p.CurrentPage > p.TotalPages {
		p.CurrentPage = p.TotalPages
	}

	// If there is only one page, don't render pagination.
	if p.TotalPages <= 1 {
		classes["pagination--hidden"] = true
	}

	pageArray := make([]int, p.TotalPages)

	for i := range pageArray {
		pageArray[i] = i
	}

	return h.Div(
		classes,
		h.DataAttr("current-page", fmt.Sprintf("%d", p.CurrentPage)),
		h.DataAttr("total-pages", fmt.Sprintf("%d", p.TotalPages)),
		h.Ul(
			h.Class("pagination__list"),
			// Chevron left
			h.Li(
				h.Class("pagination__btn pagination__btn--left"),
				h.A(
					Icon(&IconProps{
						Identifier: "chevron-left",
						Classes: c.Classes{
							"pagination__icon": true,
						},
					}),
				),
			),
			g.Group(g.Map(pageArray, func(i int) g.Node {
				page := i + 1
				classes := c.Classes{
					"pagination__item": true,
				}
				if page == p.CurrentPage {
					classes["pagination__item--current"] = true
				}
				return h.Li(
					classes,
					h.A(
						g.Text(fmt.Sprintf("%d", page)),
					),
				)
			})),
			// Chevron right
			h.Li(
				h.Class("pagination__btn pagination__btn--right"),
				h.A(
					Icon(&IconProps{
						Identifier: "chevron-right",
						Classes: c.Classes{
							"pagination__icon": true,
						},
					}),
				),
			),
		),

		InlineStyle(Assets, "/Pagination.css"),
		InlineScript(Assets, "/Pagination.js"),
	)

}

File: components/Pagination.js
(function () {
  const pagination = me();
  const paginationList = me("ul", pagination);
  const prevPageBtn = me("li.pagination__btn--left", paginationList);
  const nextPageBtn = me("li.pagination__btn--right", paginationList);

  let currentPageNumber = parseInt(pagination.dataset.currentPage);
  const totalPages = parseInt(pagination.dataset.totalPages);

  function updatePaginationState() {
    if (currentPageNumber === 1) {
      prevPageBtn.classList.add("pagination__btn--disabled");
    } else {
      prevPageBtn.classList.remove("pagination__btn--disabled");
    }

    if (currentPageNumber === totalPages) {
      nextPageBtn.classList.add("pagination__btn--disabled");
    } else {
      nextPageBtn.classList.remove("pagination__btn--disabled");
    }
  }

  function initPagination() {
    updatePaginationState();
    addUrlParams([
      { name: "page", value: currentPageNumber },
      { name: "pageSize", value: 100 },
    ]);
  }

  function goToNextPage() {
    const nextPageNumber = currentPageNumber + 1;
    const pages = paginationList.children;

    if (nextPageNumber === totalPages) {
      nextPageBtn.classList.add("pagination__btn--disabled");
    }

    if (nextPageNumber <= totalPages) {
      pages[currentPageNumber].classList.remove("pagination__item--current");
      pages[nextPageNumber].classList.add("pagination__item--current");
      currentPageNumber++;
      updatePaginationState();
      addUrlParams([
        { name: "page", value: currentPageNumber },
        { name: "pageSize", value: 100 },
      ]);
    }
  }

  function goToPrevPage() {
    const prevPageNumber = currentPageNumber - 1;
    const pages = paginationList.children;

    if (prevPageNumber === 1) {
      prevPageBtn.classList.add("pagination__btn--disabled");
    }

    if (prevPageNumber > 0) {
      pages[currentPageNumber].classList.remove("pagination__item--current");
      pages[prevPageNumber].classList.add("pagination__item--current");
      currentPageNumber--;
      updatePaginationState();
      addUrlParams([
        { name: "page", value: currentPageNumber },
        { name: "pageSize", value: 100 },
      ]);
    }
  }

  // Event Listeners
  prevPageBtn.on("click", (e) => {
    e.preventDefault();
    goToPrevPage();
  });

  nextPageBtn.on("click", (e) => {
    e.preventDefault();
    goToNextPage();
  });

  paginationList.on("click", (e) => {
    e.preventDefault();

    if (e.target.tagName === "A") {
      const pages = paginationList.children;
      const pageNumber = parseInt(e.target.textContent);

      pages[currentPageNumber].classList.remove("pagination__item--current");
      pages[pageNumber].classList.add("pagination__item--current");
      currentPageNumber = pageNumber;
      updatePaginationState();
      addUrlParams([
        { name: "page", value: currentPageNumber },
        { name: "pageSize", value: 100 },
      ]);
    }
  });

  initPagination();
})();
